src/wp-includes/meta.php

```diff
@@ -170,6 +170,123 @@ function add_metadata( $meta_type, $object_id, $meta_key, $meta_value, $unique =
 	return $mid;
 }
 
+/**
+ * Adds multiple items of metadata for the specified object.
+ *
+ * This function is more performant than calling `add_metadata()` multiple times because it queries the database only
+ * once and clears the meta cache only once.
+ *
+ * This function will always insert all of the provided metadata even if matching keys already exist. This behaviour
+ * matches that of add_metadata() when its `$unique` parameter is set to false.
+ *
+ * If the insert fails, no metadata will be inserted. It's not possible for a subset of the rows to be inserted.
+ *
+ * Examples:
+ *
+ *     bulk_add_metadata(
+ *         'post',
+ *         $post_id,
+ *         array(
+ *             'meta_key_1' => 'value_1',
+ *             'meta_key_2' => 'value_2',
```

Comment:
As there can be multiple meta data rows with the same key, should the API be?
 
```suggestion
 *             'meta_key_1' => [ 'key_1_value_1', 'key_1_value_2' ],
 *             'meta_key_2' => [ 'key_2_value_1' ],
```

Reply:
[Timothy said something similar here](https://github.com/WordPress/wordpress-develop/pull/5128#issuecomment-3225713643) but I think it complicates the API for the less common use case. Users will inevitably expect to be able to pass a string value and then we end up needing to support either the array values or non-array values.

I'm not against it, but I think it's a complication that perhaps we can do without. What do you think?

Reply:
@johnbillion I guess it's fine to keep it simple. Reviewing wp_insert_user and wp_insert_post, the same assumption is made there so there.